Improve CONTRIBUTIONS.md with more information (#372)
oamg/convert2rhel: CONTRIBUTING.md

GitHub. [GitHub Quickstart - Fork a repo](https://docs.github.com/en/get-started/quickstart/fork-a-repo)

As an additional material, check out this Red Hat blog post
about [What is an open source upstream?](https://www.redhat.com/en/blog/what-open-source-upstream)

### Collaborating with Pull Requests

Check out this guide from GitHub on how to collaborate with pull requests. This is an in-depth guide on everything you
need to know about PRs, forks, contributions and much
more. [GitHub - Collaborating with pull requests](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests)

## Getting started with development

### Dependencies for local development

We have some required dependencies you should have installed on your system (either your local computer or a docker
container) to get ready to write some code:

Required dependencies:

- [virtualenv](https://virtualenv.pypa.io/en/latest/)
- [python](https://www.python.org/downloads/)
- [pre-commit](https://pre-commit.com/)
- [git](https://git-scm.com/downloads)
- [podman](https://podman.io/getting-started/installation) or a working [docker](https://docs.docker.com/get-docker/)

Optional dependencies:

> - [make](https://www.gnu.org/software/make/#download)

Author: Yeah, I guess we can remove the links, they are not that helpful right now.

Maybe, should we add command instructions for how to install those programs in Fedora? Or even other distros...

* Limit the first line to 72 characters or fewer
* If you are fixing a GitHub issue, include something like "Closes #xyz"

To know more about best practices for commit messages, go ahead and
read [How to Write a Git Commit Message](https://chris.beams.io/posts/git-commit/).

### Python code styleguide

For Convert2RHEL we use the [PEP 8 -- Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/) as well
some linters/formatters that are run directly from [pre-commit](https://pre-commit.com).

Don't worry, most of the code formatting and styleguide is handled by [black](https://github.com/psf/black), an awesome
formatter for python code.

### Python documentation styleguide

The documentation for our python code should follow
the [PEP 8 -- Documentation Strings](https://www.python.org/dev/peps/pep-0008/#toc-entry-20)
or more precisely the [PEP 257 -- Docstring Conventions](https://www.python.org/dev/peps/pep-0257/) and preferably all
functions should contain at least a minimum docstring to identify what that function is used for.

## Additional information

### Building rpms locally

For building the rpms locally, you will need:

> - podman or a working docker

Reviewer: Implies that docker doesn't usually work :stuck_out_tongue: 
```suggestion
- podman or docker
```

```bash
make tests # This will runs all tests on CentOS7 and CentOS8 containers
```

and if you wish to run the tests on only one system, you can do so with:

```bash
make tests8 # Will run the tests only on CentOS8
make tests7 # Will run the tests only on CentOS7
```

Finally, you can find out all the unit testing for Convert2RHEL at
the [convert2rhel/convert2rhel/unit_tests](https://github.com/oamg/convert2rhel/tree/main/convert2rhel/unit_tests)
folder.

#### Writing integration tests

The general idea of our integration tests is to run the convert2rhel tool as a black box and look for explicit outputs given a system state and inputs - like environment variables and CLI parameters.

Usually, the integration tests are written by the QE team in Convert2RHEL using all the best practices they already
have for integration tests, but every developer can help them to write the integration tests as well.

If you want to know more on how to execute the integration tests locally or how to create them, you can read the
guide [Integration tests](https://github.com/oamg/convert2rhel/tree/main/tests/integration) which covers all the

> necessary steps to execute the integration tests locally.

Reviewer 2: I would add something like this:
```suggestion
necessary steps to execute the integration tests locally. The structure of the integration tests can be found in the [convert2rhel/plans](https://github.com/oamg/convert2rhel/tree/main/plans) folder. We use [fmf](https://fmf.readthedocs.io/en/stable/index.html) format to define which which tests will be executed. Tests are then executed using the [tmt](https://tmt.readthedocs.io/en/stable/index.html) tool.
Tests themself are written in mostly in `Python`. However one can write the integration test in any other language, eg. in `Bash`.
```

# Contributing to Convert2RHEL

The following is a set of guidelines for contributing to Convert2RHEL codebase, which are hosted in
the [OAMG Organization](https://github.com/oamg) on GitHub. These are mostly guidelines, not rules.

<div align="center">

> 

Author: Yeah, about that. I saw GitHub already has some TOC implemented for markdown stuff, should we go ahead and use theirs instead of this manual TOC? https://github.com/oamg/convert2rhel/pull/372#discussion_r755224423

### Table Of Contents

</div>

[What should I know before I get started?](#what-should-i-know-before-i-get-started)

* [Working with GitHub](#working-with-github)
* [Setting up Git](#setting-up-git)
* [Forking a repository](#forking-a-repository)
* [Collaborating with Pull Requests](#collaborating-with-pull-requests)

[Getting started with development](#getting-started-with-development)

* [Dependencies for local development](#dependencies-for-local-development)
* [Setting up the environment](#setting-up-the-environment)
* [Writing tests](#writing-tests)

[Styleguide](#styleguide)

* [Git commit messages](#git-commit-message)
* [Python code styleguide](#python-code-styleguide)
* [Python documentation styleguide](#python-documentation-styleguide)

[Additional Notes](#additional-information)

* [Building rpms locally](#building-rpms-locally)

> * [Building copr builds locally](#building-copr-builds-locally)

Author: GitHub already has a TOC built-in in the markdown visualization, should we keep it in the document or remove it to use the own TOC provided by GitHub? 

![image](https://user-images.githubusercontent.com/2881357/143050075-41a9631e-b747-4743-9994-2b2eadd8df49.png)

Author: @SpyTec what do you think about that?

Reviewer: Sounds good to me

Using the second rule of thumb as an example, much of the times, we don't have any control of what the package managers
API will return, so instead of depending on an ever-changing return call, we can mock it so every time we run the tests
we know we are getting the same output, over and over again.

Rule of thumb to when you should write the `functional unit tests`:

1. When you are reading a file from the system;
2. When you are doing some version comparison;
3. Everytime when you are not interacting with an unstable service (eg: Databases, APIs, PM, etc...);

Functional tests tend to be more flexible in definition than the `traditional tests`, because it will do the actual call
for the function being tested, and not just a call to a Mock class in which can return anything we desire.

Both the `traditional` and the `functional` tests always focus on the same thing. Writing test for the smallest unit
possible (eg: a function, class method, etc...)

> Beware, `functional tests` can sometimes be confused with `integration tests`, but they play a role far different.
> Functional tests are for testing a small unit of code as being "Functional", and Integration tests are for testing a

> > workflow or the whole application without prior knowing what the code does.

Reviewer: I think it would be better to rewrite this to be more of when and how to use a mock rather than splitting into two categories

Author: I've included some examples of mock/no-mock, and tried to rewrite and keep it concise...

##### Running the unit tests

Unit tests should be run inside a container for either `CentOS 7` or `CentOS 8`. We provide some `make` commands to easy
the job on running those tests.

> WARNING: Those commands will run every single unit tests we have on the containers

For running the tests in a container, you will need to execute:

```bash

> make tests # This will runs all tests on CentOS7 and CentOS8 containers

Reviewer: ```suggestion
make tests  # This runs all tests on CentOS 7 and CentOS 8 containers
```

```

and if you wish to run the tests on only one system, you can do so with:

```bash

> make tests8 # Will run the tests only on CentOS8
> make tests7 # Will run the tests only on CentOS7

Reviewer: ```suggestion
make tests8  # Will run the tests only on CentOS 8
make tests7  # Will run the tests only on CentOS 7
```

```

Finally, you can find out all the unit testing for Convert2RHEL at
the [convert2rhel/convert2rhel/unit_tests](https://github.com/oamg/convert2rhel/tree/main/convert2rhel/unit_tests)
folder.

#### Writing integration tests

> The general idea of writing `integration tests` is, as the name say, test an integration in your current application,
> and this integration can be a specific workflow, running the entire application or simple testing if our application is
> using the correct combination of CLI parameters works. Ideally, the integration tests does not know about the code (this
> is why we have unit testing), but rather, it's more concerned about `inputs` and `states`.

Reviewer: ```suggestion
The general idea of our integration tests is to run the convert2rhel tool as a black box and look for explicit outputs given a system state and inputs - like environment variables and CLI parameters.
```

Usually, the `integration tests` are written by the `QE team` in Convert2RHEL using all the best practices they already
have for integration tests, but every developer can help them to write the integration tests as well. If you have any
doubt or questions about that, [@mkluson](https://github.com/mkluson) and [@kokesak](https://github.com/kokesak) are the
right guys to reach out!

If you want to know more on how to execute the integration tests locally or how to create them, you can read the
guide [Integration tests](https://github.com/oamg/convert2rhel/tree/main/tests/integration) which covers all the
necessary steps to execute the integration tests locally.

Finally, you can find out all the integration testing for Convert2RHEL at
the [convert2rhel/tests](https://github.com/oamg/convert2rhel/tree/main/tests) folder.

## Styleguide

### Git commit message

* Use the present tense ("Add feature" not "Added feature")
* Use the imperative mood ("Move cursor to..." not "Moves cursor to...")
* Limit the first line to 72 characters or fewer

> * If you are fixing a GitHub issue, include something like "Closes issue #xyz"

Reviewer: IIRC GitHub only recognizes `Closes #xxx` and `Fixes #xxx`
```suggestion
* If you are fixing a GitHub issue, include something like "Closes #xyz"
```

To know more about best practices for commit messages, go ahead and

> read [How to Write a Git Commit Message](https://chris.beams.io/posts/git-commit/).

Author: Indeed! It's very descriptive and a good resource.

### Python code styleguide

For Convert2RHEL we use the [PEP 8 -- Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/) as well
some linters/formatters that are run directly from [pre-commit](https://pre-commit.com).

> Don't worry, most of the code formatting and styleguide is handled by [black](https://github.com/psf/black), an awesome
> tool for formatting python code.

Reviewer: ```suggestion
Don't worry, most of the code formatting and styleguide is handled by [black](https://github.com/psf/black), an awesome
formatter for python code.
```

* [Working with GitHub](#working-with-github)
* [Setting up Git](#setting-up-git)
* [Forking a repository](#forking-a-repository)
* [Collaborating with Pull Requests](#collaborating-with-pull-requests)

[Getting started with development](#getting-started-with-development)

* [Dependencies for local development](#dependencies-for-local-development)
* [Setting up the environment](#setting-up-the-environment)
* [Writing tests](#writing-tests)

[Styleguide](#styleguide)

* [Git commit messages](#git-commit-message)
* [Python code styleguide](#python-code-styleguide)
* [Python documentation styleguide](#python-documentation-styleguide)

[Additional Notes](#additional-information)

* [Building rpms locally](#building-rpms-locally)
* [Building copr builds locally](#building-copr-builds-locally)

## What should I know before I get started?

Below are a list of things to keep in mind when developing and submitting contributions to this repository.

1. All python code must be compatible with versions 2.6/2.7/3.6.
2. The code should follow linting from pylint.
3. All commits should have passed the pre-commit checks.

> 4. Always squash your commits into a single one, [how to squash commits](https://stackoverflow.com/a/3825347).

Author: I've included a section to squash commits, since we do it in every pr.

Reviewer: We don't necessarily need to squash commits if you have a good split between what each commit should do. Most of the time people squash to one commit as it makes it easier
